### PR TITLE
Deprecate compressed ids

### PR DESCRIPTION
**Pull request by imtayadeway**
Support for compressed ids in URLs, the filter, and in the request body was added and incorporated into the Fine release (see https://github.com/ManageIQ/manageiq/pull/11621,https://github.com/ManageIQ/manageiq/pull/11577, https://github.com/ManageIQ/manageiq/pull/11248, https://github.com/ManageIQ/manageiq/pull/10322).

Since that can't be removed immediately, I'm deprecating its use by maintaining the outward behavior but adding warnings to the log.

Somewhat surprisingly, the behavior of `Application.compressed_id?` was altered in https://github.com/ManageIQ/manageiq/pull/11621 to return true for uncompressed ids as well. Hence, I needed a new predicate (added in https://github.com/ManageIQ/manageiq-api/commit/8d740f6c5c4bcc5f131335dbbca7f935686dae9b) that encapsulated the original meaning of "compressed and definitely not a regular id". I chose to add it in this repository - if needed elsewhere it can be extracted. This collision of naming/concepts is a bit regrettable, but I'm OK with it as a compromise. I think the right thing to do would be to change `ApplicationRecord.compressed_id` to more accurately describe its behavior, update callers, etc., but this seems like it would be a big investment in code we are determined to be short-lived. 

/cc @chrisarcand @Fryguy 
@miq-bot add-label technical debt
@miq-bot assign @abellotti

### CODE REVIEW

File: lib/api.rb [hunk 1]
```diff
@@ -33,4 +33,17 @@ def self.url_attribute?(attr)
...
+  # similar), we need a way to distinquish compressed ids from
+  # anything else so that we can deprecate their usage.
+  def self.compressed_id?(thing)
+    ApplicationRecord::RE_COMPRESSED_ID === thing.to_s
+  end
+
+  def self.uncompress_id(id)
+    $api_log.warn("The use of compressed ids is deprecated, and the support for which will be removed in a future release.")
+    ApplicationRecord.uncompress_id(id)
+  end
```
[the remark this answers is not in the document]
Comment (chrisarcand) [reply lib/api.rb]: Prefer [`deprecate_methods`](http://api.rubyonrails.org/classes/ActiveSupport/Deprecation/MethodWrapper.html#method-i-deprecate_methods) which [our deprecations](https://github.com/manageiq/manageiq/blob/master/lib/vmdb/deprecation.rb) hook in to.

[the remark this answers is not in the document]
Comment (imtayadeway) [reply lib/api.rb]: After playing around with this, I'm not sure that `deprecate_methods` is the thing we want. We're not deprecating a method here, so much as a feature of the API. End-users may not know anything about the internals of the API, and may not know how to interpret a message that uses terminology from the implementation. Any thoughts?

[the remark this answers is not in the document]
Comment (chrisarcand) [reply lib/api.rb]: I had the same thought of your warning message just adding on to what will already be deprecated in the other API. And end users won't care about your API log :)

@abellotti @Fryguy Any thoughts here? Is there any documentation per-version for the API that should reflect this deprecation from the requester's side instead?

[the remark this answers is not in the document]
Comment (abellotti) [reply lib/api.rb]: othen than logging in the api.log the deprecation, we can definitely document this, something in line with http://manageiq.org/docs/reference/latest/api/appendices/deprecated_resource_attributes

[the remark this answers is not in the document]
Comment (imtayadeway) [reply lib/api.rb]: With that in mind, is there any point in writing to the log here?

[the remark this answers is not in the document]
Comment (abellotti) [reply lib/api.rb]: For completeness I guess, in case customers look at one and not the other.

[the remark this answers is not in the document]
Comment (chrisarcand) [reply lib/api.rb]: Yeah whatever, keep it.

**Comment by chrisarcand** [comment]
LGTM, leaving it to @abellotti for any other feedback.

**Comment by imtayadeway** [comment]
Bump @abellotti anything else needed for this (other than documenting)?

**Comment by abellotti** [comment]
LGTM!! @imtayadeway, if you could take care of the one rubocop issue, that'll be appreciated.

**Comment by imtayadeway** [comment]
@abellotti AFAIK there is no better way to return a boolean as the result of matching a regex, so I think rubocop can be ignored unless you have a better suggestion?

**Comment by chrisarcand** [comment]
@imtayadeway I do like what you did, but if you want to appease the cop...

```ruby
!!(ApplicationRecord::RE_COMPRESSED_ID =~ thing)
```

**Comment by chrisarcand** [comment]
As discussed on Gitter and [sanctioned elsewhere](https://github.com/ManageIQ/manageiq/issues/11954), the use of `!!` is appropriate. LGTM.

**Comment by abellotti** [comment]
Thank you @imtayadeway and @chrisarcand for the update.  LGTM!!

**Pull request merged**